dali/pipeline/data/tensor_vector.cc

```diff
@@ -12,268 +12,494 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include <string>
 #include "dali/pipeline/data/tensor_vector.h"
+#include <string>
+#include "dali/core/access_order.h"
 #include "dali/core/common.h"
 #include "dali/core/error_handling.h"
+#include "dali/core/tensor_layout.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/types.h"
 
 namespace dali {
 
+namespace {
+
+/**
+ * @brief Check if both shared pointers have the same managed pointer (not the one returned by
+ * .get())
+ */
+bool same_owner(const std::shared_ptr<void> &x, const std::shared_ptr<void> &y) {
+  if (x.owner_before(y) || y.owner_before(x))
+    return false;
+  return true;
+}
+
+
+bool same_owner(const std::weak_ptr<void> &x, const std::shared_ptr<void> &y) {
+  if (x.owner_before(y) || y.owner_before(x))
+    return false;
+  return true;
+}
+
+
+// TODO(klecki): move this to the class?
+TensorShape<> empty_shape(int dim) {
+  TensorShape<> result;
```

Comment:
If this is intended to produce a shape that has a volume of 0, then we should check that `dim` is not 0.
```suggestion
  assert(dim > 0);
  TensorShape<> result;
```